mindmeld/text_preparation/normalizers.py

```diff
@@ -195,28 +196,105 @@ def normalize(self, text):
+                dictionaries. The example rule below removes any text in parentheses.
+                {
+                    "description": "remove_text_in_parantheses",
+                    "pattern": "\(.+?\)",
+                    "replacement": ""
+                }
+        """
+        super().__init__()
+        self.regex_norm_rules = regex_norm_rules
+
+    def normalize(self, text):
+        """Apply Regex substitutions to the given text.
+        Args:
+            text (str): Input text.
+        Returns:
+            normalized_text (str): Normalized Text.
+        """
+        # TODO: Handle case if the "pattern" or "replacement" is none
+
+        for regex_norm_rule in self.regex_norm_rules:
+            pattern = regex_norm_rule.get("pattern")
+            replacement = regex_norm_rule.get("replacement")
+            text = re.sub(pattern, replacement, text)
+        return text
+
+
+class MindMeldRegexNormalizer(RegexNormalizer):
+    """Normalization class that substitutes regex matches with either an empty string or a
+    replacement string."""
+
+    def __init__(self):
+        """Creates a Regex Normalizer instance.
+
+        Args:
+            regex_norm_rules (List[Dict]): Regex normalization rules represented as
+                dictionaries. The example rule below removes any text in parentheses.
+                {
+                    "description": "remove_text_in_parantheses",
+                    "pattern": "\(.+?\)",
+                    "replacement": ""
+                }
+        """
+        super().__init__(DEFAULT_REGEX_NORM_RULES)
+
+
 class NormalizerFactory:
     """Normalizer Factory Class"""
 
     @staticmethod
-    def get_normalizer(normalizer):
+    def get_normalizer(normalizer, regex_norm_rules=None):
         """A static method to get a Normalizer
 
         Args:
             normalizer (str): Name of the desired Normalizer class
+            regex_norm_rules (List[Dict], optional): Regex normalization rules represented as
+                dictionaries. The example rule below removes any text in parentheses.
+                {
+                    "description": "remove_text_in_parantheses",
+                    "pattern": "\(.+?\)",
+                    "replacement": ""
+                }
         Returns:
             (Normalizer): Normalizer Class
         """
-        if normalizer == NoOpNormalizer.__name__:
-            return NoOpNormalizer()
-        elif normalizer == ASCIIFold.__name__:
-            return ASCIIFold()
-        elif normalizer == NFC.__name__:
-            return NFC()
-        elif normalizer == NFD.__name__:
-            return NFD()
-        elif normalizer == NFKC.__name__:
-            return NFKC()
-        elif normalizer == NFKD.__name__:
-            return NFKD()
-        raise AssertionError(f" {normalizer} is not a valid Normalizer.")
+        normalizer_classes = [
```

Comment:
This is a perfect use case for maps! If you use a callable for the value, you don't need the special case for `RegexNormalizer`.

```python
normalizer_classes = {
    NoOpNormalizer.__name__: NoOpNormalzer,
    ...,
    RegexNormalizer.__name__: lambda: RegexNormalzer(regex_norm_rules),
}
cls = normalizer_classes.get(normalizer)
if not cls:
    raise TypeError('invalid normalizer type')
return cls()
```

Reply:
Thanks for the advice! Went ahead and applied this strategy for all components of the text_preparation_pipeline